Parker bi ds physio

dpark6060 opened the pull request:
New class for physio data.  Handles:

- Conversion of trigger times to binary timeseries'
- duplicate tic times of consecutive samples
- Processing data, accounting for irregular sampling rates (multiple options for how to handle, including ignore) as well as different start times
- QA reports that compare raw data to processed data
- BIDSification 
- enhanced robustness

dpark6060 committed:
Merge pull request #8 from flywheel-apps/master

Master to BIDS

dpark6060 committed:
added to do in run.py

dpark6060 committed:
First Attempt at setting file properties

dpark6060 committed:
HUGE update for ECG stuff.  File Classification work present but commented (in dev)

dpark6060 committed:
added ECG QC (Untested)

dpark6060 committed:
Complete rework.  Physio dict is now in object that does all the
functions.  It's super cool.

dpark6060 committed:
Removed redundant functions

dpark6060 committed:
Debugging for Docker Run

dpark6060 committed:
Passed first gear local tests

dpark6060 committed:
Added Doc strings plus successful instance run

dpark6060 committed:
Updated Gitignore for local files

dpark6060 committed:
Removed Unused imports

joshicola commented on Dockerfile:
```diff
@@ -28,6 +28,8 @@ RUN apt-get -qq update && apt-get -qq install -y \
+
 RUN pip3 install -r requirements.txt
 
 RUN rm -rf /root/.cache/pip && \
```
Dockerfile Protip:
remove desired files in the layer that created them

RUN pip3 install -r requirements.txt &&  rm -rf /root/.cache/pip 

This will reduce the actual size of the layer.

I just learnt this..... :(

joshicola replied:
Dockerfile Protip:
remove desired files in the layer that created them

RUN pip3 install -r requirements.txt && rm -rf /root/.cache/pip

This will reduce the actual size of the layer.

I just learnt this..... :(

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
+
+        # Catch any exceptions
+        except Exception as e:
+            gear_context.log.fatal(traceback.print_exc())
```
`gear_context.log.exception(e)` 
will print the exception and the traceback

joshicola commented on run.py:
```diff
@@ -22,522 +20,169 @@
 ##--------  Gear Specific files/folders/values  ----##
 tic_len = 2.5e-3  # seconds, length of one "tick"
```
not used for the rest of the script. ??

joshicola commented on run.py:
```diff
@@ -22,522 +20,169 @@
+    custom dicts object.  Each physio dict has a key "['bids_file']" and "['log_file'].  if these files were created,
+    these keys point to that filename.  Loop through the log files and the bids files and set metadata based off
+    dictionary parameters and inhereted properties.
+    :param context: gear context
```
sphinx documentation needs to have types as well:

    :param context: The gear context with core functionality
    :type context: class: `flywheel.gear_context.GearContext`

joshicola commented on run.py:
```diff
@@ -1,5 +1,5 @@
-
+import traceback
 import flywheel
 from utils import futils as fu
```
NICE!
There is something about well-named variables!!! 
"FU" works very well ;)

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
+
+            # Extract matches keyword for BIDS data format
+            matches = gear_context.get_input('DICOM_ARCHIVE')['object']['info']
+            if 'SeriesDescription' in matches.keys() and matches['SeriesDescription']:
```
This should create an error if 'SeriesDescription' is not in matches.keys().

`if 'SeriesDescription' in matches.keys():`
?

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
+
+                # If BIDS is desired:
+                if gear_context.config['Generate_Bids']:
+                    physio.bids_o_matic_9000(processed=process, matches=matches, zip_output=False)
```
bwahahahahahaha!!!!
bids-o-matic_9000!!!
south park much?

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
+                gear_context.log.info('Removing .log files')
+                cmd = ['/bin/rm', output_dir,'*.log']
+
+                exec_command(cmd)
```
This version of the `exec_command()` takes the gear context as one of the inputs.

joshicola commented on run.py:
```diff
@@ -22,522 +20,169 @@
+        classification['Custom'] =['Physio']
+    except:
+        context.log.info('  Cannot determine classification from config.json.')
+        classification = {'Custom':['Physio']}
```
50: classification starts out as a list ([])
53: classification key ('Custom') is assigned to ['Physio'] (ERROR!!!)
56: classification is assigned to a dictionary 

53 & 56 seem redundant to one another... whereas  50 contradicts 53???

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
-        gear_context.log.debug('Successfully generated physio dict')
+            # First let's check the output.  We'll try to make this robust for any measurement:
+            gear_context.log.debug('Globbing Physio')
+            all_physio = glob.glob(op.join(physio_output_dir, '*.log'))
+            info_file = glob.glob(op.join(physio_output_dir, '*Info.log'))
```
making these files distinct could avoid L292 altogether... but not much of a "save" in any regard.

joshicola commented on run.py:
```diff
@@ -553,67 +198,147 @@ def main():
+                    continue
+                physio_objects.append(phys.physio(phys_file))
+
+            for physio in physio_objects:
```
Is there some reason for keeping these loops (L289 & L296) distinct?

joshicola commented on run.py:
```diff
@@ -22,522 +20,169 @@
 ##--------  Gear Specific files/folders/values  ----##
 tic_len = 2.5e-3  # seconds, length of one "tick"
 
+def data_classifier(context,output_dir):
```
Where is this function ever used?  I found it defined... but not used. ???

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+        The less efficient but easier to follow method to process data.  Equivalent to "one_step_process_raw", which
+        is faster.  Steps:
+        1. remove duplicate tics from the raw data
+        2. crate a new tic array to handle missing data based on the users "interp" and "handle missing data" settings
```
Is this "create" rather than "crate"?

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+            raise Exception("Error converting triggers to timeseries for {}".format(self.type)) from e
+
+
+    def process_raw(self):
```
"Readability Rules!!!"

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+            self.proc_data = True
+
+        except Exception:
+            raise
```
Give a message here?  Or leave it as is?

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+        """
+        This function is only going to ensure that the channels have the same length by padding at the sample rate
+        This is to help BIDS output.
+        :return:
```
If there is no "return value" this is unnecessary.

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+        """
+        pl.figure()
+        for chan in self.channels:
+            pl.plot(self.acq_tics[chan],self.chan_values[chan])
```
Did you want legends here?
I don't see these referenced anywhere....

joshicola commented on utils/physio.py:
```diff
@@ -0,0 +1,620 @@
+
+
+
+
```
Overall,
I like the integration of your QC explorations into classes directly associated with each physio object.

joshicola reviewed (changes requested):
There are some parts of the code that will raise unintended exceptions if not addressed.

AndysWorth commented on Dockerfile:
```diff
@@ -28,6 +28,8 @@ RUN apt-get -qq update && apt-get -qq install -y \
 
```
Should "flywheel/python:dcmtk.latest" (up above) have a version number instead of "latest"?

AndysWorth commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
 flywheel-sdk
```
this should have a version number.  Maybe some others should also be pinned to a version?

AndysWorth reviewed (approved):
Looks great to me!

dpark6060 committed:
Update metadata for BIDS

dpark6060 committed:
Update metadata for BIDS

dpark6060 committed:
Dev1 commit for Bridge Center

dpark6060 committed:
Dev2 commit for Bridge Center

dpark6060 committed:
Dev2 commit for Bridge Center, fixed NaN output issue.

dpark6060 committed:
Physio file metadata test

dpark6060 committed:
Physio file metadata test

dpark6060 committed:
passed physio metadata test for BIDS data and physio labels.  Added
more extensive docstrings

dpark6060 committed:
Final changes from last code review
- added versions in dockerfile (Andy)
- added types to docstrings
- attempted to improve exception calling?
- yeah...
